### PR TITLE
TL 54: Add Tenure Shared Code

### PR DESCRIPTION
**Pull request by notatiyyah**
## Link to JIRA ticket

https://hackney.atlassian.net/browse/TL-54

## Describe this PR

- Add the Tenure domain objects and tests.
- Update the pipeline to run the tests and check code formatting for each push.

#### _Checklist_

- [x] Added tests to cover all new production code
- [x] Checked all code for possible refactoring
- [x] Code pipeline builds correctly

### *Follow up actions after merging PR*

Update the Tenure API and Tenure Listener to use this package.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -0,0 +1,32 @@
+.PHONY: setup
```
Comment (mikemoody-amido) [inline Makefile:1]: This file is not needed.

File: Hackney.Shared.Tenure/Sns/User.cs [hunk 2]
```diff
@@ -0,0 +1,8 @@
+namespace Hackney.Shared.Tenure.Sns
+{
+    public class User
+    {
+        public string Name { get; set; }
+        public string Email { get; set; }
+    }
+}
```
Comment (mikemoody-amido) [inline Hackney.Shared.Tenure/Sns/User.cs:8]: Not needed here. The Sns event object is common but should not go in this package.

File: Hackney.Shared.Tenure/Sns/TenureSns.cs [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+
+        public DateTime DateTime { get; set; }
+
+        public Guid EntityId { get; set; }
+
+        public EventData EventData { get; set; }
+
+        public User User { get; set; }
+    }
+}
```
Comment (mikemoody-amido) [inline Hackney.Shared.Tenure/Sns/TenureSns.cs:27]: Not needed here. The Sns event object is common but should not go in this package.

File: Hackney.Shared.Tenure/Sns/EventData.cs [hunk 4]
```diff
@@ -0,0 +1,8 @@
+namespace Hackney.Shared.Tenure.Sns
+{
+    public class EventData
+    {
+        public object OldData { get; set; }
+        public object NewData { get; set; }
+    }
+}
```
Comment (mikemoody-amido) [inline Hackney.Shared.Tenure/Sns/EventData.cs:8]: Not needed here. The Sns event object is common but should not go in this package.

File: Hackney.Shared.Tenure/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,18 @@
+FROM mcr.microsoft.com/dotnet/core/sdk:3.1
+
+ARG LBHPACKAGESTOKEN
+ENV LBHPACKAGESTOKEN=$LBHPACKAGESTOKEN
```
Comment (mikemoody-amido) [inline Hackney.Shared.Tenure/Dockerfile:4]: This is only need if you're using the Hackney.Core packages.
So you can either remove it or 
* Undelete the NuGet.Config file and update it to look like one of the other Api projects.
* Add this line after the other COPY commands (l.9)
```COPY /NuGet.Config /root/.nuget/NuGet/NuGet.Config```
* Ensure the CircleCi pipeline has the LBHPACKAGESTOKEN env variable set.

**Comment by mikemoody-amido** [comment]
Forgot to add a general comment on my review...

* Some stuff to remove as it shouldn't be there.
* I would keep/update the NuGet setup so that Hackney.Core packages are accessible if needed. 
* You've just taken the Domain models here, but should we also consider the response models and/or the Db models (and the factory conversions between them) as well? And the validation?
@amidomichael A question for you I think?

File: Hackney.Shared.Tenure.Tests/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,24 @@
+FROM mcr.microsoft.com/dotnet/core/sdk:3.1
+
+# disable microsoft telematry
+ENV DOTNET_CLI_TELEMETRY_OPTOUT='true'
+
+ENV DynamoDb_LocalMode='true'
+ARG LBHPACKAGESTOKEN
+ENV LBHPACKAGESTOKEN=$LBHPACKAGESTOKEN
```
[the remark this answers is not in the document]
Comment (notatiyyah) [reply Hackney.Shared.Tenure.Tests/Dockerfile:7]: Removed the NuGet config as it was causing errors in the pipeline but ran fine locally. All fixed now, though!

File: docker-compose.yml [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+    ports:
+      - 3000:3000
+    environment:
+     - DynamoDb_LocalMode=true
+     - DynamoDb_LocalServiceUrl=http://dynamodb-database:8000
+     - AWS_REGION=eu-west-2
+     - AWS_ACCESS_KEY_ID=local
+     - AWS_SECRET_ACCESS_KEY=local
+    links:
+      - dynamodb-database
```
[the remark this answers is not in the document]
Comment (notatiyyah) [reply docker-compose.yml]: If this isn't needed, can't I delete the entire dockerfile (`Hackney.Shared.Tenure/Dockerfile`)?

File: nuget.config [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+        <clear />
+        <add key="github" value="https://api.nuget.org/v3/index.json" protocolVersion="3" />
+    </packageSources>
+    <packageSourceCredentials>
+        <github>
+            <add key="Username" value="%USERNAME%" />
+            <add key="ClearTextPassword" value= "%NUGET_KEY%" />
+        </github>
+    </packageSourceCredentials>
+</configuration>
```
Comment (mikemoody-amido) [inline nuget.config:13]: You need to update this so it looks like [one from one of the Apis](https://github.com/LBHackney-IT/tenure-information-api/blob/master/nuget.config)...

File: docker-compose.yml (hunk 7 above)
[the remark this answers is not in the document]
Comment (mikemoody-amido) [reply docker-compose.yml]: Yes

**Review by humulla — APPROVED** [review]

**Pull request merged**